File: 3.js
var lengthOfLongestSubstring = function (s) {
    let start = 0;
    let max = 0;
    let chars = {};
    for (let end = 0; end < s.length; end++) {
        let current = s[end];
        if (chars[current] !== undefined) {
            start = Math.max(chars[current] + 1, start);
        }
        max = Math.max(end - start + 1, max);
        chars[current] = end;
    }
    return max;
};

File: 3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        start = 0
        longest = 0
        chars = {}
        for end, ch in enumerate(s):
            if ch in chars:
                start = max(chars[ch] + 1, start)
            longest = max(end - start + 1, longest)
            chars[ch] = end
        return longest

# time complexity: O(N) - iterate through string
# space complexity: O(N) - at most N unique chars within dict
